modules/Blockchain/Ethereum/contracts/Marketplace.sol

```diff
@@ -0,0 +1,260 @@
+        require(uint256(sellerIdentity) != 0, "Cannot initiate purchase without seller id submitted");
+        require(uint256(buyerIdentity) != 0, "Cannot initiate purchase without price submitted");
+        require(uint256(originalDataRootHash) !=0, "Cannot initiate purchase without originalDataRootHash submitted");
+        require(uint256(encodedDataRootHash) != 0, "Cannot initiate purchase without encodedDataRootHash submitted");
+
+        require(ERC725(buyerIdentity).keyHasPurpose(keccak256(abi.encodePacked(msg.sender)), 2) || ERC725(buyerIdentity).keyHasPurpose(keccak256(abi.encodePacked(msg.sender)), 1), "Sender does not have proper permission to call this function!");
+
+        // purchase_id is created as a hash of: sender, previous block hash, and data root hash
+        bytes32 purchaseId = keccak256(abi.encodePacked(msg.sender, blockhash(block.number - 1), encodedDataRootHash));
+
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setSeller(purchaseId, sellerIdentity);
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setBuyer(purchaseId, buyerIdentity);
+
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setPrice(purchaseId, price);
+
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setOriginalDataRootHash(purchaseId, originalDataRootHash);
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setEncodedDataRootHash(purchaseId, encodedDataRootHash);
+
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setTimestamp(purchaseId, block.timestamp);
+        MarketplaceStorage(hub.getContractAddress("MarketplaceStorage")).setStage(purchaseId, 1);
+
+        // Secure funds from buyer
+        reserveTokens(buyerIdentity, price);
+
+        emit PurchaseInitiated(purchaseId, sellerIdentity, buyerIdentity, price, encodedDataRootHash);
+    }
+
+    function revealKey(bytes32 purchaseId, bytes32 key) public {
+        require(uint256(purchaseId) != 0, "Cannot reveal key without purchase id submitted");
+        require(uint256(key) != 0, "Cannot reveal key without key submitted");
+
+        MarketplaceStorage marketplaceStorage = MarketplaceStorage(hub.getContractAddress("MarketplaceStorage"));
+
+        address sellerIdentity = marketplaceStorage.getSeller(purchaseId);
+        require(ERC725(sellerIdentity).keyHasPurpose(keccak256(abi.encodePacked(msg.sender)), 2) || ERC725(sellerIdentity).keyHasPurpose(keccak256(abi.encodePacked(msg.sender)), 1), "Only seller can reveal key!");
+        require(marketplaceStorage.getStage(purchaseId) == 1, "Key can only be revealed in the initiated stage");
+
+        marketplaceStorage.setKey(purchaseId, key);
+
+        marketplaceStorage.setTimestamp(purchaseId, block.timestamp);
+        marketplaceStorage.setStage(purchaseId, 2);
+
+        // Secure funds from all parties
+        reserveTokens(sellerIdentity, marketplaceStorage.getPrice(purchaseId));
```

Comment:
Do we reserve tokens here for a potential refund by seller?

Reply:
We reserve tokens in case there was tampering with the data, then the buyer gets refunded with the tokens reserved here